src/config_file.c

```diff
@@ -1915,36 +1254,30 @@ static int config_write(diskfile_backend *cfg, const char *key, const regex_t *p
 	int result;
 	char *section, *name, *ldot;
 	git_filebuf file = GIT_FILEBUF_INIT;
-	git_buf buf = GIT_BUF_INIT;
-	struct reader reader;
+	git_buf buf = GIT_BUF_INIT, contents = GIT_BUF_INIT;
+	git_config_parser reader;
 	struct write_data write_data;
 
 	memset(&reader, 0, sizeof(reader));
-	git_buf_init(&reader.buffer, 0);
 	reader.file = &cfg->file;
 
 	if (cfg->locked) {
-		result = git_buf_puts(&reader.buffer, git_buf_cstr(&cfg->locked_content));
+		result = git_buf_puts(&contents, git_buf_cstr(&cfg->locked_content));
 	} else {
 		/* Lock the file */
 		if ((result = git_filebuf_open(
 			     &file, cfg->file.path, GIT_FILEBUF_HASH_CONTENTS, GIT_CONFIG_FILE_MODE)) < 0) {
-			git_buf_free(&reader.buffer);
+			git_buf_free(&contents);
 			return result;
 		}
 
 		/* We need to read in our own config file */
-		result = git_futils_readbuffer(&reader.buffer, cfg->file.path);
+		result = git_futils_readbuffer(&contents, cfg->file.path);
 	}
 
 	/* Initialise the reading position */
-	if (result == GIT_ENOTFOUND) {
-		reader.read_ptr = NULL;
-		reader.eof = 1;
-		git_buf_clear(&reader.buffer);
-	} else if (result == 0) {
-		reader.read_ptr = reader.buffer.ptr;
-		reader.eof = 0;
+	if (result == 0 || result == GIT_ENOTFOUND) {
+		git_parse_ctx_init(&reader.ctx, contents.ptr, contents.size);
```

Comment:
Do we guarantee that we won't touch the fields in `contents` if it's not found? It makes sense that we wouldn't touch them, but if that's not part of the contract, this could initialise the context with whatever random pointer.

Reply:
Do you mean the contents of the context?